### PR TITLE
[Merged by Bors] - feat: port Data.ZMod.Basic

### PR DESCRIPTION
**Pull request by ChrisHughes24**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/ZMod/Basic.lean [hunk 1]
```diff
@@ -167,76 +173,83 @@ theorem cast_eq_val [NeZero n] (a : ZMod n) : (a : R) = a.val := by
...
 @[norm_cast]
-theorem int_cast_zMod_cast (a : ZMod n) : ((a : ℤ) : ZMod n) = a := by
+theorem int_cast_zmod_cast (a : ZMod n) : ((a : ℤ) : ZMod n) = a := by
   cases n
-  · rw [Int.cast_id a, Int.cast_id a]
-  · rw [coe_coe, Int.cast_ofNat, Fin.cast_val_eq_self]
-#align zmod.int_cast_zmod_cast ZMod.int_cast_zMod_cast
+  · simp [ZMod.cast, ZMod]
+  · dsimp [ZMod.cast, ZMod]
+    erw [Int.cast_ofNat, Fin.cast_val_eq_self]
```
Comment (eric-wieser) [inline Mathlib/Data/ZMod/Basic.lean]: Any idea why the rewrites stopped working?

Comment (ChrisHughes24) [reply Mathlib/Data/ZMod/Basic.lean]: Something to do with needing the fact that `ZMod (n+1) = Fin (n+1)`. The new `rw` unfolds a little less aggressively I guess.

File: Mathlib/Data/ZMod/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,1211 @@
...
+/-- The identity between `ZMod m` and `ZMod n` when `m = n`, as a ring isomorphism. -/
+def ringEquivCongr {m n : ℕ} (h : m = n) : ZMod m ≃+* ZMod n := by
+  cases' m with m <;> cases' n with n
+  · exact RingEquiv.refl _
+  · exfalso
+    exact n.succ_ne_zero h.symm
+  · exfalso
+    exact m.succ_ne_zero h
+  ·
+    exact
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/ZMod/Basic.lean:442]:
```suggestion
  · exact
```

File: Mathlib/Data/ZMod/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,1211 @@
...
+      {
+        Fin.cast
+          h with
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/ZMod/Basic.lean:445]:
```suggestion
      { Fin.cast h with
```

File: Mathlib/Data/ZMod/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,1211 @@
...
+
+theorem valMinAbs_nonneg_iff {n : ℕ} [NeZero n] (x : ZMod n) : 0 ≤ x.valMinAbs ↔ x.val ≤ n / 2 := by
+  rw [valMinAbs_def_pos]; split_ifs with h
+  · exact iff_of_true (Nat.cast_nonneg _) h
+  · exact iff_of_false (sub_lt_zero.2 <| Int.ofNat_lt.2 x.val_lt).not_le h
+#align zmod.val_min_abs_nonneg_iff ZMod.valMinAbs_nonneg_iff
+
+theorem valMinAbs_mul_two_eq_iff {n : ℕ} (a : ZMod n) : a.valMinAbs * 2 = n ↔ 2 * a.val = n := by
+  cases' n with n; · simp
+  by_cases a.val ≤ n.succ / 2
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/ZMod/Basic.lean:953]:
```suggestion
  cases' n with n
  · simp
  by_cases h : a.val ≤ n.succ / 2
```

File: Mathlib/Data/ZMod/Basic.lean [hunk 5]
```diff
@@ -0,0 +1,1211 @@
...
+
+@[simp]
+theorem valMinAbs_zero : ∀ n, (0 : ZMod n).valMinAbs = 0
+  | 0 => by simp only [valMinAbs_def_zero]
+  | n + 1 => by simp only [valMinAbs_def_pos, if_true, Int.ofNat_zero, zero_le, val_zero]
+#align zmod.val_min_abs_zero ZMod.valMinAbs_zero
+
+@[simp]
+theorem valMinAbs_eq_zero {n : ℕ} (x : ZMod n) : x.valMinAbs = 0 ↔ x = 0 := by
+  cases' n with n; · simp
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/ZMod/Basic.lean:1007]:
```suggestion
  cases' n with n
  · simp
```

File: Mathlib/Data/ZMod/Basic.lean [hunk 6]
```diff
@@ -0,0 +1,1211 @@
...
+  rw [valMinAbs_def_pos]
+  split_ifs
+  · rw [Int.natAbs_ofNat, nat_cast_zmod_val]
+  · rw [← Int.cast_ofNat, Int.ofNat_natAbs_of_nonpos this, Int.cast_neg, Int.cast_sub,
+      Int.cast_ofNat, Int.cast_ofNat, nat_cast_self, sub_zero, nat_cast_zmod_val]
+#align zmod.nat_cast_nat_abs_val_min_abs ZMod.nat_cast_natAbs_valMinAbs
+
+theorem valMinAbs_neg_of_ne_half {n : ℕ} {a : ZMod n} (ha : 2 * a.val ≠ n) :
+    (-a).valMinAbs = -a.valMinAbs := by
+  cases' eq_zero_or_neZero n with h h;
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/ZMod/Basic.lean:1026]:
```suggestion
  cases' eq_zero_or_neZero n with h h
```

File: Mathlib/Data/ZMod/Basic.lean [hunk 7]
```diff
@@ -0,0 +1,1211 @@
...
+  rw [← add_le_add_iff_right x.natAbs]
+  refine' le_trans (le_trans ((add_le_add_iff_left _).2 hl) _) (Int.natAbs_sub_le _ _)
+  rw [add_sub_cancel, Int.natAbs_mul, Int.natAbs_ofNat]
+  refine' le_trans _ (Nat.le_mul_of_pos_right <| Int.natAbs_pos.2 hm)
+  rw [← mul_two]; apply Nat.div_mul_le_self
+#align zmod.nat_abs_min_of_le_div_two ZMod.natAbs_min_of_le_div_two
+
+theorem natAbs_valMinAbs_add_le {n : ℕ} (a b : ZMod n) :
+    (a + b).valMinAbs.natAbs ≤ (a.valMinAbs + b.valMinAbs).natAbs := by
+  cases' n with n; · rfl
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/ZMod/Basic.lean:1092]:
```suggestion
  cases' n with n
  · rfl
```

File: Mathlib/Data/ZMod/Basic.lean [hunk 8]
```diff
@@ -0,0 +1,1211 @@
...
+
+private theorem mul_inv_cancel_aux (a : ZMod p) (h : a ≠ 0) : a * a⁻¹ = 1 := by
+  obtain ⟨k, rfl⟩ := nat_cast_zmod_surjective a
+  apply coe_mul_inv_eq_one
+  apply Nat.coprime.symm
+  rwa [Nat.Prime.coprime_iff_not_dvd Fact.out, ← CharP.cast_eq_zero_iff (ZMod p)]
+
+/-- Field structure on `ZMod p` if `p` is prime. -/
+instance : Field (ZMod p) :=
+  { inferInstanceAs (CommRing (ZMod p)),inferInstanceAs (Inv (ZMod p)),
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/ZMod/Basic.lean:1108]:
```suggestion
  { inferInstanceAs (CommRing (ZMod p)), inferInstanceAs (Inv (ZMod p)),
```

**Review by Ruben-VandeVelde — CHANGES_REQUESTED** [review]
lgtm otherwise

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**